cpp/source/deal.II/geometry.templates.h

```diff
@@ -30,47 +134,250 @@ void Geometry<dim>::fill_materials_map(
             material_database.get<std::string>("material_id"));
     std::string const material_name =
         material_database.get<std::string>("name");
-    materials->emplace(material_name, material_ids);
+    _materials->emplace(material_name, material_ids);
   }
 }
 
 template <int dim>
+void Geometry<dim>::fill_materials_map()
+{
+  _materials = std::make_shared<std::unordered_map<
+      std::string, std::vector<dealii::types::material_id>>>();
+  (*_materials)["anode"] = std::vector<dealii::types::material_id>(0, 0);
+  (*_materials)["separator"] = std::vector<dealii::types::material_id>(1, 1);
+  (*_materials)["cathode"] = std::vector<dealii::types::material_id>(1, 2);
+  (*_materials)["collector_anode"] =
+      std::vector<dealii::types::material_id>(1, 3);
+  (*_materials)["collector_cathode"] =
+      std::vector<dealii::types::material_id>(1, 4);
+  std::vector<dealii::types::material_id> coll_material_ids(2);
+  coll_material_ids[0] = 3;
+  coll_material_ids[1] = 4;
+  (*_materials)["collector"] = coll_material_ids;
+}
+
+template <int dim>
 Geometry<dim>::Geometry(
     std::shared_ptr<boost::property_tree::ptree const> database,
     boost::mpi::communicator mpi_communicator)
-    : mpi_communicator(mpi_communicator)
+    : _communicator(mpi_communicator)
 {
-  triangulation = std::make_shared<dealii::distributed::Triangulation<dim>>(
+  _triangulation = std::make_shared<dealii::distributed::Triangulation<dim>>(
       mpi_communicator);
-  std::string const mesh_file = database->get<std::string>("mesh_file");
-  dealii::GridIn<dim> mesh_reader;
-  mesh_reader.attach_triangulation(*triangulation);
-  std::fstream fin;
-  fin.open(mesh_file.c_str(), std::fstream::in);
-  std::string const file_extension =
-      mesh_file.substr(mesh_file.find_last_of(".") + 1);
-  if (file_extension.compare("ucd") == 0)
+  // If mesh_file does not exist, we need generate the mesh.
+  if (boost::optional<std::string> mesh_file =
+          database->get_optional<std::string>("mesh_file"))
   {
-    mesh_reader.read_ucd(fin);
+    dealii::GridIn<dim> mesh_reader;
+    mesh_reader.attach_triangulation(*_triangulation);
+    std::fstream fin;
+    fin.open(mesh_file->c_str(), std::fstream::in);
+    std::string const file_extension =
+        mesh_file->substr(mesh_file->find_last_of(".") + 1);
+    if (file_extension.compare("ucd") == 0)
+    {
+      fill_materials_map(database);
+      boost::property_tree::ptree boundary_database =
+          database->get_child("boundary_values");
```

Comment:
Most of it yes but I thought we still need to know how to map the boundary id from the mesh that we read to the cathode and the anode boundary id.